in-toto-spec.md

```diff
@@ -736,56 +736,79 @@ The `ARTIFACT_RULE` format is the following:
   this step.
-* **CREATE**: Indicates that products matched by the pattern must not appear as
-  materials of this step. Note, the rule still passes if the pattern does not
-match any products
+* **CREATE**: indicates that products matched by the pattern must not appear as
+  materials of this step.
 * **DELETE**: indicates that materials matched by the pattern must not appear
-  as products of this step. Note, the rule still passes if the pattern does not
-match any materials.
-* **MODIFY**: indicates that the PATH's hash must be updated in this step. In
-  other words, this path must appear as a material and a product within this
-step and their hashes must not match.
-
-The artifact rules contained in the `"expected_materials"` and
-`"expected_products"` fields operate in a similar fashion as firewall rules do.
-This means that the first rule that matches a specific artifact in the link
-metadata will be used to match that artifact. In addition, there is an implicit
-`ALLOW *` at the end of such fields.
-
-##### 4.3.3.1 MATCH rule behavior
-
-The `MATCH` rule is used to tie different steps together, by means of their
-materials and products. The main rationale behind the MATCH rule is to identify
-the sources of artifacts as they are passed around in the supply chain. In this
-sense, the MATCH rule will be used to identify which step should be providing a
+  as products of this step.
+* **MODIFY**: indicates that products matched by this pattern must appear as
+  materials of this step, and their hashes must not by the same.
+
+
+##### 4.3.3.1 Rule processing
+
+Artifact rules reside in the `"expected_products"` and `"expected_materials"`
+fields of a step and are applied sequentially on a queue of `"materials"` or
+`"products"` from the step's corresponding link metadata. They operate in a
+similar fashion as firewall rules do. This means if an artifact is successfully
+consumed by a rule, it is removed from the queue and cannot be consumed by
+subsequent rules. There is an implicit `"ALLOW *"` at the end of each rule
+list. By explicitly specifying `"DISALLOW *"`, in-toto verification fails if an
+artifact was not consumed by an earlier rule. Here, we describe an algorithm to
+illustrate the behavior of the rules being applied:
+
+```python
+VERIFY_EXPECTED_ARTIFACTS(rule_set, link, target_links)
+
+# load the artifacts from the link
+artifacts = load_artifacts_as_queue(link)
+
+# iterate over all the rules
+for rule in rules:
+  consumed_artifacts, rule_error = apply_rule(rule, artifacts)
+
+  if rule_error:
+    return ERROR("Rule failed to verify!")
+
+  artifacts -= consumed_artifacts
+
+return SUCCESS
+```
+
+
+##### 4.3.3.2 MATCH rule behavior
+
+The match rule is used to tie different steps together, by means of their
+materials and products. The main rationale behind the match rule is to identify
+the provenience of artifacts as they are passed around in the supply chain. In this
```

Comment:
I see what you tried to do by s/source/provenience/, although I'm not sure it's the word that we want. Any ideas?

Reply:
Not sure either about provenience either. I replaced "sources" because it made me think of source code, but maybe that's just me. I can either revert or change it to "origin"?

Reply:
[Here's an example](https://github.com/in-toto/docs/blame/master/in-toto-spec.md#L1530) in the spec, where "sources" is used to refer to source code files.

Reply:
Yeah, that's why I said `I see what you tried to do`. I'm not sure, maybe origin works better...

Reply:
This shouldn't be a blocker. I do slightly prefer origins over sources, but I will change provenience (back) to whatever you think is better. :)

Reply:
Rather than "provenience", use "provenance", it's the more established term.

Reply:
Thanks for the comment, @reza-curtmola. Santiago and I decided that we'll change it to "origins". :)